src/services/pin-state-transition/hooks/prepare-activity-log.js

```diff
@@ -0,0 +1,140 @@
+const mongoose = require('mongoose');
+const errors = require('feathers-errors');
+const Pin = require('../../pin/pin-model');
+const actions = require('../../../constants/actions');
+const states = require('../../../constants/pin-states');
+
+const safetyCheck = (hook) => {
+  // hook.params.user should be populated by auth.populateUser before hook
+  if (!hook.params.user) {
+    throw new errors.GeneralError('Internal error: User is not populated');
+  }
+
+  // hook.params.pinId should be provided via request URL
+  if (!hook.params.pinId || !mongoose.Types.ObjectId.isValid(hook.params.pinId)) {
+    throw new errors.NotFound(`No pin found for id '${hook.params.pinId}'`);
+  }
+
+  // hook.data.state should be provided via body data
+  if (!hook.data.state) {
```

Comment:
For `create` method, `hook.data` is always an object. (https://github.com/feathersjs/feathers-commons/blob/8e6b85681a2f1c6f8193c781cddb8086c21fba6d/src/hooks.js#L27-L33)